src/cls/rgw/cls_rgw.cc

```diff
@@ -583,7 +588,9 @@ int rgw_bucket_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
 
   bool done = false;   // whether we need to keep calling get_obj_vals
   bool more = true;    // output parameter of get_obj_vals
-  bool has_delimiter = !op.delimiter.empty();
+  const bool has_delimiter = !op.delimiter.empty();
+  const bool rgw_do_osd_side_checks =
+    cls_get_config(hctx)->rgw_do_osd_side_bucket_list_checks;
 
   if (has_delimiter &&
```

Comment:
Hmm, perhaps the straight case (no delimiter, no extensive versioning) of bucket listing could bypass the `cls_rgw` entirely – RGW (or maybe even `cls_rgw_client`) could emit `CEPH_OSD_OP_OMAPGETVALS`  (instead of `CEPH_OSD_OP_CALL`) relying the `is_valid()` and `is_visible()` checks are performed client-side.

The question is how large the _ugly namespace_ of keys can be.

Dynamic switching between the direct and cls-assisted paths is not unthinkable.